plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -747,24 +754,35 @@ private function get_final_frontend_query( $query, $page = 1, $is_exclude_applie
 	 * @param bool  $is_exclude_applied_filters Whether to exclude the applied filters or not.
 	 */
 	private function get_final_query_args( $common_query_values, $query, $is_exclude_applied_filters = false ) {
-		$handpicked_products = $query['handpicked_products'] ?? array();
-		$orderby_query       = $query['orderby'] ? $this->get_custom_orderby_query( $query['orderby'] ) : array();
-		$on_sale_query       = $this->get_on_sale_products_query( $query['on_sale'] );
-		$stock_query         = $this->get_stock_status_query( $query['stock_status'] );
-		$visibility_query    = is_array( $query['stock_status'] ) ? $this->get_product_visibility_query( $stock_query, $query['stock_status'] ) : array();
-		$featured_query      = $this->get_featured_query( $query['featured'] ?? false );
-		$attributes_query    = $this->get_product_attributes_query( $query['product_attributes'] );
-		$taxonomies_query    = $query['taxonomies_query'] ?? array();
-		$tax_query           = $this->merge_tax_queries( $visibility_query, $attributes_query, $taxonomies_query, $featured_query );
-		$date_query          = $this->get_date_query( $query['timeFrame'] ?? array() );
-		$price_query_args    = $this->get_price_range_query_args( $query['priceRange'] ?? array() );
+		$orderby_query    = $query['orderby'] ? $this->get_custom_orderby_query( $query['orderby'] ) : array();
+		$on_sale_query    = $this->get_on_sale_products_query( $query['on_sale'] );
+		$stock_query      = $this->get_stock_status_query( $query['stock_status'] );
+		$visibility_query = is_array( $query['stock_status'] ) ? $this->get_product_visibility_query( $stock_query, $query['stock_status'] ) : array();
+		$featured_query   = $this->get_featured_query( $query['featured'] ?? false );
+		$attributes_query = $this->get_product_attributes_query( $query['product_attributes'] );
+		$taxonomies_query = $query['taxonomies_query'] ?? array();
+		$tax_query        = $this->merge_tax_queries( $visibility_query, $attributes_query, $taxonomies_query, $featured_query );
+		$date_query       = $this->get_date_query( $query['timeFrame'] ?? array() );
+		$price_query_args = $this->get_price_range_query_args( $query['priceRange'] ?? array() );
 
 		// We exclude applied filters to generate product ids for the filter blocks.
 		$applied_filters_query = $is_exclude_applied_filters ? array() : $this->get_queries_by_applied_filters();
 
-		$merged_query = $this->merge_queries( $common_query_values, $orderby_query, $on_sale_query, $stock_query, $tax_query, $applied_filters_query, $date_query, $price_query_args );
+		$merged_query = $this->merge_queries(
+			$common_query_values,
+			$orderby_query,
+			$on_sale_query,
+			$stock_query,
+			$tax_query,
+			$applied_filters_query,
+			$date_query,
+			$price_query_args
+		);
+
+		$handpicked_products = $query['handpicked_products'] ?? array();
+		$related_products    = $this->get_related_products( $query['related_to'], $common_query_values['posts_per_page'] );
```

Comment:
One concern I have about using `'posts_per_page'` is that additional filters would receive only `n` products and filter to less, even if the related products would be more.

Should we perhaps set this to something higher (or unlimited?) and then let the other filtering take place? This could work, however, some sites have an unreasonably large number of products and that might not be great in terms of performance.

Reply:
Hmm Good question! It's valid concern about the potential limitations of using 'posts_per_page' in this context. Thinking out loud:
- As you suggested, We can try setting it to a bit higher value (e.g. 200) and then let the other filtering take place. This way, we won't have to worry about performance issues.
	- But it's really hard to find a value that would work in all cases. Ideal solution would be to have `wc_get_related_products` to return all related products. By fetching all related products, we can ensure that subsequent filters (like price range or attributes) have the full set of related products to work with.
- Maybe we can try finding alternative solution which doesn't include using `wc_get_related_products` to get all related products?
- Once we call `wc_get_related_products`, Maybe we can use filters which allow us to modify the query itself to make it more performant?
	- AFAIU `wc_get_related_products` internally end up calling [get_related_products](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php#L1460) function. This function have `woocommerce_product_related_posts_query` filter. Maybe we can use this filter to modify the query to make it more performant?
- Maybe we can check if it's possible to use [posts_clauses](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L121) to update the final WP_Query, similar to how we are doing for Price Range filter as you can see in [add_price_range_filter_posts_clauses](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L1570) function. 

It feels like a tough problem to solve, but we need to find a solution to make it work for all sites. Maybe we can start with easier approach and then move towards more complex approach. Let me know your thoughts 🙇‍♂️ 

CC: @kmanijak In case you have other suggestions 😅

Reply:
> Maybe we can try finding alternative solution which doesn't include using `wc_get_related_products` to get all related products?

The benefit of using `wc_get_related_products()` is that it's the standard way to get other products related to a product. It has lots of filters already in place and I'm sure there's plenty of extensions that modify the behavior to suit their individual project. Ideally any solution we come up with should use this same functionality so that there's parity with existing "Related Products" behavior.

> Once we call `wc_get_related_products`, Maybe we can use filters which allow us to modify the query itself to make it more performant?

I'm not _sure_ what kind of performance characteristics related products have. I just know it's a second query that runs to get a list of products to include.

> It feels like a tough problem to solve, but we need to find a solution to make it work for all sites. Maybe we can start with easier approach and then move towards more complex approach. Let me know your thoughts 🙇‍♂️

I'd like for us to approach this more conservatively at first. We should keep using `wc_get_related_products()` and make any improvements that we need to later on when they become apparent. I don't really know what kind of performance implications this might have and it's probably something we should test on a large store. The main concern I have here is that it's a second query (get related product IDs and use them in the main query) but I'd wager the two query approach is faster than something with a bunch of convoluted joins to find IDs that exist in a subset etc etc etc

For what it's worth, there _is_ a cache in place for related products already so all of this is probably fine :smile: To the original point, let's see what removing the limit entirely does and how that impacts the overall performance of the collection.

Reply:
> Ideally any solution we come up with should use this same functionality so that there's parity with existing "Related Products" behavior.

Aha! Good point. Makes sense to me. 

> I'm not sure what kind of performance characteristics related products have. I just know it's a second query that runs to get a list of products to include.

Maybe I am overthinking it 🤷🏻 
- I thought calling `wc_get_related_products` with large limit will have performance implications. 
- Also, Having lot of ID's in `post__in` will have performance implications, as we know that `post__in` is processed using IN clause in SQL. 

But as you said, we can probably improve it later if needed.

> To the original point, let's see what removing the limit entirely does and how that impacts the overall performance of the collection.

Sounds good to me. We can always improve it later if needed 😀

Reply:
> We should keep using wc_get_related_products() and make any improvements that we need to later on when they become apparent.

I'm also in favor of this.

> For what it's worth, there is a cache in place for related products already so all of this is probably fine

I believe it's a good idea to set a limit here, even if it affects the query builder. 

Looking at the code in `wc_get_related_products()`, it seems that the transients [are written](https://github.com/woocommerce/woocommerce/blob/d2fb60a5d12705cc984a79151c1a284b2af96a52/plugins/woocommerce/includes/wc-product-functions.php#L999) _regardless_ of the limit, and each limit request generates a new one if there are "not enough cached". 

As of now, every store is flooded with transients that contain _five_ ids, yes. If we opt for a limitless query, considering a large catalog with thousands of IDs in each transient, it could potentially affect resources(?) It's definitely a tough problem. I suggest using a reasonable limit, not 5, but not unlimited. Perhaps 100 would be a good compromise?

Reply:
I've gone with 100 in my latest push. We can always change the number later and I think the point @xristos3490 is a _really_ good one. There are plenty of sites with way more products than you can imagine and that would be wild to potentially cache every permutation of relationship between products.